File: sdk/typescript/src/workspace.ts
import { OrgoClient } from './client.js';
import { Computer } from './computer.js';
import type { CreateComputerOpts } from './types.js';

export class Workspace {
  readonly id: string;
  name: string;
  private _client: OrgoClient;

  constructor(client: OrgoClient, data: Record<string, unknown>) {
    this._client = client;
    this.id = data.id as string;
    this.name = (data.name as string) || '';
  }

  static async create(client: OrgoClient, name: string): Promise<Workspace> {
    const data = await client.post<{ workspace: Record<string, unknown> }>('/workspaces', { name });
    return new Workspace(client, data.workspace);
  }

  static async list(client: OrgoClient): Promise<Workspace[]> {
    const data = await client.get<{ workspaces: Record<string, unknown>[] }>('/workspaces');
    return data.workspaces.map(ws => new Workspace(client, ws));
  }

  async get(): Promise<this> {
    const data = await this._client.get<{ workspace: Record<string, unknown> }>(`/workspaces/${this.id}`);
    this.name = data.workspace.name as string;
    return this;
  }

  async update(name: string): Promise<void> {
    await this._client.patch(`/workspaces/${this.id}`, { name });
    this.name = name;
  }

  async delete(opts?: { force?: boolean }): Promise<void> {
    const qs = opts?.force ? '?force=true' : '';
    await this._client.delete(`/workspaces/${this.id}${qs}`);
  }

  async computers(): Promise<Computer[]> {
    const data = await this._client.get<{ computers: Record<string, unknown>[] }>(`/workspaces/${this.id}/computers`);
    return data.computers.map(c => new Computer(this._client, c));
  }

  async createComputer(opts?: Partial<CreateComputerOpts>): Promise<Computer> {
    const data = await this._client.post<{ computer: Record<string, unknown> }>(
      `/workspaces/${this.id}/computers`,
      { name: opts?.name || 'computer', cpu: opts?.cpu || 2, ram: opts?.ram || 4096, resolution: opts?.resolution },
    );
    return new Computer(this._client, data.computer);
  }
}

File: sdk/python/src/orgo_clone/workspace.py
from __future__ import annotations
from typing import Any, TYPE_CHECKING
from .client import OrgoClient

if TYPE_CHECKING:
    from .computer import Computer


class Workspace:
    """Represents a workspace containing computers."""

    def __init__(self, client: OrgoClient, data: dict[str, Any]) -> None:
        self._client = client
        self.id: str = data["id"]
        self.name: str = data.get("name", "")

    @classmethod
    async def create(cls, client: OrgoClient, name: str) -> "Workspace":
        data = await client.post("/workspaces", json={"name": name})
        return cls(client, data["workspace"])

    @classmethod
    async def list(cls, client: OrgoClient) -> list["Workspace"]:
        data = await client.get("/workspaces")
        return [cls(client, ws) for ws in data["workspaces"]]

    async def get(self) -> "Workspace":
        data = await self._client.get(f"/workspaces/{self.id}")
        self.name = data["workspace"]["name"]
        return self

    async def update(self, name: str) -> None:
        await self._client.patch(f"/workspaces/{self.id}", json={"name": name})
        self.name = name

    async def delete(self, force: bool = False) -> None:
        params = {"force": "true"} if force else {}
        await self._client.delete(f"/workspaces/{self.id}", params=params)

    async def computers(self) -> list["Computer"]:
        from .computer import Computer
        data = await self._client.get(f"/workspaces/{self.id}/computers")
        return [Computer(self._client, c) for c in data["computers"]]

    async def create_computer(self, **kwargs: Any) -> "Computer":
        from .computer import Computer
        data = await self._client.post(f"/workspaces/{self.id}/computers", json=kwargs)
        return Computer(self._client, data["computer"])
